[UXUI] Action to block domains from submitting in form results

andersonjeccel opened the pull request:
| Q                                      | A
| -------------------------------------- | --------------------------
| Bug fix?          | ❌ <!-- Use emojis to indicate positive or negative for each item. -->
| New feature/enhancement?      | ✔️
| Deprecations?                          | ❌
| BC breaks?        | ❌
| Automated tests included?              | ❌ 
| Related user documentation PR URL      | mautic/user-documentation#...
| Related developer documentation PR URL | mautic/developer-documentation-new#... 
| Issue(s) addressed                     | <!-- prefix each issue number with "Fixes #" -->

## Description

This PR implements an action to block domains from submitting in form results. The functionality provides buttons that populate the configuration field "Do not accept submission from these domain names" with domain names extracted from email fields in form submissions.



<!--
Please write a short README, 1 sentence, for your feature/bugfix. This will help people understand your PR and what it aims to do. If you are fixing a bug and if there is no linked issue already, please provide steps to reproduce the issue here.
-->
<!--
| Before                                 | After
| -------------------------------------- | ---
|                                        | 
-->


---
### 📋 Steps to test this PR:

<!--
This part is crucial. Take the time to write very clear, annotated and step by step test instructions, because testers may not be developers.
-->
1. Open this PR on Gitpod or pull down for testing locally (see docs on testing PRs [here](mdc:https:/contribute.mautic.org/contributing-to-mautic/tester))
2. Navigate to Forms → [Select a form] → Results
3. Look for form results that contain email addresses - you should see a "Block domain submissions" button for each result with an email
4. Click the "Block domain submissions" button on an individual result
5. Verify that a success message appears confirming the domain has been blocked
6. Check the batch operations section - you should see a "Block domains from submitting" button
7. Select multiple form results using the checkboxes
8. Click the "Block domains from submitting" button
9. Confirm the action in the confirmation dialog
10. Verify that all selected domains are blocked successfully
11. Go to Configuration → Forms settings and verify that the blocked domains appear in the "Do not accept submission from these domain names" field
16. Test that new submissions from blocked domains are properly rejected

<!--
If you have any deprecations and backwards compatibility breaks, list them here along with the new alternative.
-->

<!-- visual-review-evidence -->
## Visual review

The bulk action is available from form results and confirms which domain was blocked.

![Block domains action and confirmation](https://github.com/user-attachments/assets/757284a8-b163-4fd7-8468-75d1e72655e9)

The blocked domain is also visible in the form configuration.

![Blocked domain in form configuration](https://github.com/user-attachments/assets/ab6feaaa-f44b-483c-8f12-30eac4231ca2)

andersonjeccel committed:
Enable domain name submission filter by default for email fields

andersonjeccel committed:
Add domain blocking functionality to form results

andersonjeccel committed:
Apply PHP CS Fixer to FormFieldEmailType.php

andersonjeccel committed:
Merge branch 'feature/enable-email-domain-filter-by-default' into feature/Block-domain-submissions-functionality-in-forms

andersonjeccel committed:
Apply code quality improvements to FormBundle: fix PHPStan errors, apply Rector improvements, and run PHP CS Fixer

kingsedem commented:
@andersonjeccel, I have tested  ONLY steps 1 to 10 as shown below.
**"Block domains from submitting" button is available on form results**

 **Success message appears confirming the domain has been blocked**
<img width="851" height="392" alt="image" src="https://github.com/user-attachments/assets/757284a8-b163-4fd7-8468-75d1e72655e9" />

<img width="868" height="311" alt="image" src="https://github.com/user-attachments/assets/88145b83-6832-4327-b733-b460efd2e4f2" />

 For multiple form results
<img width="821" height="281" alt="image" src="https://github.com/user-attachments/assets/111bb289-ab7e-4343-9ba4-5bb354c30960" />

andersonjeccel commented:
@kingsedem I made a mistake, it should be Form settings

Thank you for pointing, just fixed the PR instructions

kingsedem commented:
ok @andersonjeccel  I just tested and I can see the blocked domain

<img width="1564" height="814" alt="image" src="https://github.com/user-attachments/assets/ab6feaaa-f44b-483c-8f12-30eac4231ca2" />

andersonjeccel committed:
Improve message

andersonjeccel committed:
Immediate validation

andersonjeccel committed:
Enable toggle only if not configured

andersonjeccel committed:
tooltip

andersonjeccel committed:
csfix

kingsedem reviewed (approved):
Tested. It works as expected

<img width="635" height="269" alt="image" src="https://github.com/user-attachments/assets/5c87c4e3-07e2-4391-982c-80c0405563ed" />

andersonjeccel committed:
Merge branch '7.x' into feature/Block-domain-submissions-functionality-in-forms

andersonjeccel committed:
Merge remote-tracking branch 'upstream/7.x' into feature/Block-domain-submissions-functionality-in-forms

andersonjeccel committed:
test coverage

andersonjeccel committed:
csfix

escopecz commented on app/bundles/FormBundle/Controller/ResultController.php:
```diff
@@ -392,6 +394,218 @@ public function deleteAction(Request $request)
+
+        if (Request::METHOD_POST === $request->getMethod()) {
+            $model = $this->getModel('form.submission');
+            \assert($model instanceof SubmissionModel);
```
Please use dependency injection
```suggestion
    public function markSpamAction(Request $request, Configurator $configurator, SubmissionModel $model): \Symfony\Component\HttpFoundation\RedirectResponse|Response
    {
        $formId   = $request->get('formId', 0);
        $objectId = $request->get('objectId', 0);
        $page     = $request->getSession()->get('mautic.formresult.'.$formId.'.page', 1);
        $flashes  = [];

        if (Request::METHOD_POST === $request->getMethod()) {
```

escopecz commented on app/bundles/FormBundle/Controller/ResultController.php:
```diff
@@ -392,6 +394,218 @@ public function deleteAction(Request $request)
+            if ('email' === $field->getType()) {
+                $alias = $field->getAlias();
+                if (!empty($results[$alias]) && str_contains($results[$alias], '@')) {
+                    return strtolower(substr(strrchr($results[$alias], '@'), 1));
```
Can you please extract this line to the EmailAddressHelper to something like `EmailAddressHelper::getDomain($emailAddress);`? It would be much more readable and it would be reusable. And easily testable.

escopecz commented on app/bundles/FormBundle/Controller/ResultController.php:
```diff
@@ -392,6 +394,218 @@ public function deleteAction(Request $request)
+        );
+    }
+
+    private function getDomainFromSubmission(Submission $submission): ?string
```
I'd renamed it to 
```suggestion
    private function getEmailDomainFromSubmission(Submission $submission): ?string
```
As I wouldn't understand this is searching for an email field in the form result and is getting the domain from the email address.

I'd also move this method inside the Submission entity. There is no dependency needed and it definitely doesn't belong to the Controller layer.

escopecz commented on app/bundles/FormBundle/Controller/ResultController.php:
```diff
@@ -392,6 +394,218 @@ public function deleteAction(Request $request)
+
+        if (Request::METHOD_POST === $request->getMethod()) {
+            $model = $this->getModel('form.submission');
+            \assert($model instanceof SubmissionModel);
```
DI
```suggestion
    public function batchMarkSpamAction(Request $request, Configurator $configurator, SubmissionModel $model): \Symfony\Component\HttpFoundation\RedirectResponse|Response
    {
        $formId  = $this->getFormIdFromRequest();
        $page    = $request->getSession()->get('mautic.formresult.'.$formId.'.page', 1);
        $flashes = [];

        if (Request::METHOD_POST === $request->getMethod()) {
```

There is also a lot of duplication between the the batch and single action. Perhaps it could be extracted to a private method and be used in both?

escopecz commented on app/bundles/FormBundle/Controller/ResultController.php:
```diff
@@ -392,6 +394,218 @@ public function deleteAction(Request $request)
+    private function enableDonotSubmitValidationOnUnconfiguredForms(): void
+    {
+        /** @var FormModel $formModel */
+        $formModel  = $this->getModel('form.form');
```
I know  Symfony won't fill in the dependency for a private method, but the right way how to do it is to send the service as a parameter and get it from the actual controller actions and from DI of those actions.
```suggestion
    private function enableDonotSubmitValidationOnUnconfiguredForms(FormModel $formModel): void
    {
```

escopecz commented on app/bundles/FormBundle/Controller/ResultController.php:
```diff
@@ -392,6 +394,218 @@ public function deleteAction(Request $request)
+             WHERE ff.type = ? 
+             AND (ff.validation IS NULL OR ff.validation = ? OR ff.validation = ?)',
+            ['email', '[]', '{}']
+        )->fetchAllAssociative();
```
Never do SQL query from a controller. This is breaking the MVC pattern and it's telling us that this method actually belongs to the Model or Repository layer. Please move it or at least the query part.

escopecz reviewed (changes requested):
Do you think you could write functional tests rather than unit tests? The unit tests doesn't ensure the feature works now and continue to work in the years to come. And they are also hard to understand and maintain.

andersonjeccel committed:
Address feedback

andersonjeccel committed:
Merge branch '7.x' into feature/Block-domain-submissions-functionality-in-forms

andersonjeccel committed:
phpunit

andersonjeccel committed:
merge 7.x into block domain submissions

andersonjeccel committed:
Merge branch '7.x' into feature/Block-domain-submissions-functionality-in-forms

andersonjeccel committed:
address form result sonar issues

andersonjeccel committed:
store blocked form domains without wildcard

andersonjeccel committed:
Merge branch '7.x' into feature/Block-domain-submissions-functionality-in-forms

andersonjeccel committed:
fix block domain test

andersonjeccel committed:
fix block domain spam test

andersonjeccel committed:
Merge branch '7.x' into feature/Block-domain-submissions-functionality-in-forms

andersonjeccel committed:
fix block domain spam validation

andersonjeccel committed:
fix form validation sonar issues

andersonjeccel committed:
Merge branch '7.x' into feature/Block-domain-submissions-functionality-in-forms

escopecz commented on app/bundles/FormBundle/Controller/ResultController.php:
```diff
@@ -392,6 +395,167 @@ public function deleteAction(Request $request)
+            return false;
+        }
+
+        $configurator->mergeParameters(['do_not_submit_emails' => $mergedDomains]);
```
This writes directly to the global do_not_submit_emails setting via Configurator::mergeParameters() and Configurator::write(), but it only gates the action with form edit permissions on the submission. Global config edits are explicitly admin-only in ConfigController.php:27. That means a non-admin user who can edit a form result can now persist application-wide configuration changes and block domains across every form. From a security standpoint, that is a privilege escalation.

andersonjeccel replied:
@escopecz  From UX standpoint, if a marketer with form access (non-admin role, but form edit permissions) wants to block something, they wouldn't be able because it's restricted to admins

I think that rewriting the feature to allow admins only would make the feature a bit weird, do you see any paths to solve this?

escopecz commented on app/bundles/FormBundle/Controller/ResultController.php:
```diff
@@ -392,6 +395,167 @@ public function deleteAction(Request $request)
+        return true;
+    }
+
+    private function enableDonotSubmitValidationOnUnconfiguredForms(FormModel $formModel): void
+    {
+        $fieldsToUpdate = $formModel->findEmailFieldsWithMissingDonotSubmitValidation();
+        $formModel->enableDonotSubmitValidationOnEmailFields($fieldsToUpdate);
+    }
```
This fans out into bulk updates in FormModel.php. Those methods scan the entire form_fields table, flip validation on every email field missing the flag, and regenerate all affected forms, with no per-form ownership or admin check. So a user acting on one submission can silently modify unrelated forms they do not control. Even if the global deny-list write were considered acceptable, this mass update crosses a second authorization boundary and broadens the blast radius from one action to the whole instance.

andersonjeccel replied:
Problem is that user expects that blocked domains start working for all forms. They would do the action once, and other forms wouldn't work as expected, that blocked address will continue submitting and the feature will look like it doesn't work.

Would you consider this a solution?: 
1. backend check all forms this user has access, and 
2. update only those + 
3. notifies something like "all forms you have access had their validation enabled so this blocked domain works for all existing forms".

escopecz replied:
I'm confused about this whole feature TBH. But updating global config values from a form result controller without checking admin permissions is a no go. And updating all forms from the same controller no matter if the user have permission to edit or even view the forms is also wrong.

I would suggest not to update global config from anywhere else except the global config form. And only update forms that the user have permission to edit.

andersonjeccel committed:
fix domain permissions

andersonjeccel committed:
merge 7.x

# Conflicts:
#	app/bundles/FormBundle/Controller/ResultController.php
#	app/bundles/FormBundle/Form/Type/FormFieldEmailType.php

andersonjeccel committed:
fix merge conflicts

andersonjeccel committed:
fix rector
